pkg/porter/archive.go

```diff
@@ -170,26 +165,131 @@ func (ex *exporter) export() error {
 	return err
 }
 
-// chtimes updates all paths under the provided archive path with a constant
-// atime and mtime (Unix time 0), such that the shasum of the resulting archive
-// will not change between repeated archival executions using the same bundle.
-// See: https://unix.stackexchange.com/questions/346789/compressing-two-identical-folders-give-different-result
-func (ex *exporter) chtimes(path string) error {
-	err := filepath.Walk(path,
-		func(subpath string, info os.FileInfo, err error) error {
+func (ex *exporter) createTarHeader(ctx context.Context, path string, file string, fileInfo os.FileInfo) (*tar.Header, error) {
+	log := tracing.LoggerFromContext(ctx)
+
+	header := &tar.Header{
+		ModTime:    time.Unix(0, 0),
+		AccessTime: time.Unix(0, 0),
+		ChangeTime: time.Unix(0, 0),
+		Uid:        0,
+		Gid:        0,
+	}
+
+	switch {
+	case fileInfo.Mode().IsDir():
+		header.Typeflag = tar.TypeDir
+		header.Mode = 0755
+	case fileInfo.Mode().IsRegular():
+		header.Typeflag = tar.TypeReg
+		header.Mode = 0644
+		header.Size = fileInfo.Size()
+	default:
+		log.Debugf("Skipping %s. Not a file/dir", file)
+		return nil, nil
+	}
+
+	// ensure header has relative file path prepended with '.'
+	relativeFilePathName := file
+
+	if filepath.IsAbs(path) {
+		relativePath, err := filepath.Rel(path, file)
+
+		if err != nil {
+			return nil, err
+		}
+
+		if relativePath != "." {
+			relativeFilePathName = fmt.Sprintf(".%s%s", string(filepath.Separator), relativePath)
+		} else {
+			relativeFilePathName = relativePath
+		}
+	}
+
+	log.Debugf("relativeFilePathName: %s\n", relativeFilePathName)
+
+	header.Name = filepath.ToSlash(relativeFilePathName)
+
+	// directories must be suffixed with '/'
+	if fileInfo.Mode().IsDir() && !strings.HasSuffix(header.Name, "/") {
+		header.Name += "/"
+	}
+
+	log.Debugf("header.Name: %s\n", header.Name)
+
+	return header, nil
+}
+
+func (ex *exporter) CustomTar(ctx context.Context, srcPath string) (io.ReadCloser, error) {
+	pipeReader, pipeWriter := io.Pipe()
+
+	gzipWriter := gzip.NewWriter(pipeWriter)
+	tarWriter := tar.NewWriter(gzipWriter)
+
+	path := filepath.Clean(filepath.Join(srcPath, "."))
```

Comment:
Can you add a comment about what this line is doing? The significance of joining "." is a bit lost on me.

Reply:
It'll remove any trailing `os.PathSeparator` from `srcPath` - I kept it as I don't how if the result from `aferox.TempDir` is consistent across the supported OS'es. It does not appear to be necessary with the "magic" join on windows nor linux.

If I can verify the consistency on a darwin machine I can remove the join - otherwise I'll add a comment stating why it's there.

Reply:
I ran the tests on my mac and what you have now in the PR works fine

Reply:
Sorry, didn't write that - I also confirmed the functionality on a mac and thus removed the magic join-with-dot code 😄